### PR TITLE
🌱 Log diffs for Cluster topology rollouts/patches

### PR DESCRIPTION
**Pull request by sbueringer**
Signed-off-by: Stefan Büringer buringerst@vmware.com

**What this PR does / why we need it**:
Follow-up of this PR https://github.com/kubernetes-sigs/cluster-api/pull/10628 for Cluster topology (also similar to #10688)

Some example logs

<details>

Setting a field in KubeadmControlPlane

> I0528 10:14:56.989054      96 reconcile_state.go:353] "Patching KubeadmControlPlane/development-10871-98gvp, diff: {\"spec\":{\"kubeadmConfigSpec\":{\"clusterConfiguration\":{\"imageRepository\":\"abc\"}}}}" controller="topology/cluster" controllerGroup="cluster.x-k8s.io" controllerKind="Cluster" Cluster="default/development-10871" namespace="default" name="development-10871" reconcileID="f6b52e3b-7ab8-48b5-a661-166d6945f726" resource={"group":"controlplane.cluster.x-k8s.io","version":"v1beta1","resource":"KubeadmControlPlane"} KubeadmControlPlane="default/development-10871-98gvp"

Rotating a DockerMachineTemplate

Rotate the template

> I0528 10:21:12.163575      96 reconcile_state.go:691] "Rotating DockerMachineTemplate/development-10871-md-0-gczmp, new name development-10871-md-0-d2kbm, diff: {\"spec\":{\"template\":{\"spec\":{\"customImage\":\"abc2\"}}}}" controller="topology/cluster" controllerGroup="cluster.x-k8s.io" controllerKind="Cluster" Cluster="default/development-10871" namespace="default" name="development-10871" reconcileID="2685e1d2-7bf8-4a81-a191-18067b159da2" MachineDeployment="default/development-10871-md-0-h72rd" machineDeploymentTopology="md-0" resource={"group":"infrastructure.cluster.x-k8s.io","version":"v1beta1","resource":"DockerMachineTemplate"} DockerMachineTemplate="default/development-10871-md-0-gczmp"

Patch MachineDeployment

> I0528 10:21:16.201107      96 reconcile_state.go:763] "Patching MachineDeployment/development-10871-md-0-h72rd, diff: {\"spec\":{\"template\":{\"spec\":{\"infrastructureRef\":{\"name\":\"development-10871-md-0-d2kbm\"}}}}}" controller="topology/cluster" controllerGroup="cluster.x-k8s.io" controllerKind="Cluster" Cluster="default/development-10871" namespace="default" name="development-10871" reconcileID="2685e1d2-7bf8-4a81-a191-18067b159da2" MachineDeployment="default/development-10871-md-0-h72rd" machineDeploymentTopology="md-0" resource={"group":"cluster.x-k8s.io","version":"v1beta1","resource":"MachineDeployment"} MachineDeployment="default/development-10871-md-0-h72rd"

MachineDeployment controller creates new MachineSet (this is from: #10688)

> I0528 10:21:16.230154      96 machinedeployment_sync.go:175] "Creating new MachineSet, couldn't find MachineSet matching MachineDeployment spec template: MachineSet development-10871-md-0-h72rd-8pt2f: diff: &v1beta1.MachineTemplateSpec{\n    ObjectMeta: {},\n    Spec: v1beta1.MachineSpec{\n      ClusterName: \"development-10871\",\n      Bootstrap:   {ConfigRef: &{Kind: \"KubeadmConfigTemplate\", Namespace: \"default\", Name: \"development-10871-md-0-xvf2n\", APIVersion: \"bootstrap.cluster.x-k8s.io\", ...}},\n      InfrastructureRef: v1.ObjectReference{\n        Kind:       \"DockerMachineTemplate\",\n        Namespace:  \"default\",\n-       Name:       \"development-10871-md-0-zm7np\",\n+       Name:       \"development-10871-md-0-d2kbm\",\n        UID:        \"\",\n        APIVersion: \"infrastructure.cluster.x-k8s.io\",\n        ... // 2 identical fields\n      },\n      Version:    &\"v1.30.0\",\n      ProviderID: nil,\n      ... // 4 identical fields\n    },\n  },MachineSet development-10871-md-0-h72rd-pw246: diff: &v1beta1.MachineTemplateSpec{\n    ObjectMeta: {},\n    Spec: v1beta1.MachineSpec{\n      ClusterName: \"development-10871\",\n      Bootstrap:   {ConfigRef: &{Kind: \"KubeadmConfigTemplate\", Namespace: \"default\", Name: \"development-10871-md-0-xvf2n\", APIVersion: \"bootstrap.cluster.x-k8s.io\", ...}},\n      InfrastructureRef: v1.ObjectReference{\n        Kind:       \"DockerMachineTemplate\",\n        Namespace:  \"default\",\n-       Name:       \"development-10871-md-0-gczmp\",\n+       Name:       \"development-10871-md-0-d2kbm\",\n        UID:        \"\",\n        APIVersion: \"infrastructure.cluster.x-k8s.io\",\n        ... // 2 identical fields\n      },\n      Version:    &\"v1.30.0\",\n      ProviderID: nil,\n      ... // 4 identical fields\n    },\n  }" controller="machinedeployment" controllerGroup="cluster.x-k8s.io" controllerKind="MachineDeployment" MachineDeployment="default/development-10871-md-0-h72rd" namespace="default" name="development-10871-md-0-h72rd" reconcileID="6d096578-2592-425c-9809-702abf427f05" Cluster="default/development-10871" MachineSet="default/development-10871-md-0-h72rd-ntptx"

</details>

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Related to https://github.com/kubernetes-sigs/cluster-api/issues/8186

### CODE REVIEW

File: internal/controllers/topology/cluster/reconcile_state.go [hunk 1]
```diff
@@ -450,7 +450,12 @@ func (r *Reconciler) reconcileCluster(ctx context.Context, s *scope.Scope) error
 		return nil
 	}
 
-	log.Infof("Patching %s", tlog.KObj{Obj: s.Current.Cluster})
+	changes := patchHelper.Changes()
```
Comment (enxebre) [inline internal/controllers/topology/cluster/reconcile_state.go:453]: has there been consideration to let this changes log be part of the patchHelper.Patch func?

File: internal/controllers/topology/cluster/reconcile_state.go [hunk 2]
```diff
@@ -751,7 +756,12 @@ func (r *Reconciler) updateMachineDeployment(ctx context.Context, s *scope.Scope
 		return nil
 	}
 
-	log.Infof("Patching %s", tlog.KObj{Obj: currentMD.Object})
+	changes := patchHelper.Changes()
+	if len(changes) == 0 {
+		log.Infof("Patching %s", tlog.KObj{Obj: currentMD.Object})
+	} else {
+		log.Infof("Patching %s, diff: %s", tlog.KObj{Obj: currentMD.Object}, changes)
+	}
```
Comment (fabriziopandini) [inline internal/controllers/topology/cluster/reconcile_state.go:764]: What if we merge with the if above
```suggestion
	if !patchHelper.HasChanges() {
		log.V(3).Infof("No changes for %s", tlog.KObj{Obj: currentMD.Object})
		return nil
	} else {
		log.Infof("Patching %s, diff: %s", tlog.KObj{Obj: currentMD.Object}, patchHelper.Changes())
	}
```

Comment (sbueringer) [reply internal/controllers/topology/cluster/reconcile_state.go:764]: I think I would still like to handle both cases (changes being empty or not). We have cases where HasChanges returns true, but Changes are empty. Technically this should only ever happen on create calls.

But to be honest I think it's brittle to depend on this exact behavior of the patch helper today. I don't want our logs to degrade if we introduce another case in the patch helper where HasChanges is true, but Changes returns an empty byte array

**Comment by enxebre** [comment]
/hold cancel
Thanks, I'm good proceeding as it is

**Pull request merged**